### ISSUE TITLE
Log4j1.x and SLF4j compatibility jars in classpath need to be included

### ISSUE THREAD

**Issue by talevy** [elastic/logstash#5781]
work in https://github.com/elastic/logstash/issues/5650 leads to a mismatch between the popular slf4j and log4j1.x logging libraries. It should be the responsibility of the plugin to include adapters if it knows it needs them. Logstash could potentially include them at runtime for all, but this feels wasteful since they will only be used against a select set of plugins.

re discussion: https://github.com/elastic/logstash/pull/5651#discussion_r75302743

**Comment by fbaligand** [comment]
Here's my opinion about it :
- in logstash, logging system and its configuration is common and shared with everybody : core code and plugins (contrary to an applications server for example)
- manage this shared logging system and its integration is logstash's responsibility, and not each plugin's responsibility.
- if it's not done, some plugin logs will be lost, some others will arrive anywhere.
- in particular for slf4j, it's important that slf4j "api" and "impl" versions are aligned. And logstash-core is the best place to ensure that.

=> That's why I think it is relevant to add these JARs to logtash-core :
- slf4j-api
- log4j-slf4j-impl
- log4j-1.2-api
- and optionally (if some plugins/libraries use Java Util Logging framework) : log4j-jul

@ph @suyograo @jordansissel 
Any opinion on that subject ?

**Comment by jordansissel** [comment]
My feeling is that we should onlyship log4j2 because we need it. It feels
like we can't solve the problem for other libraries. For example, what if a
plugin or dependency needs a conflicting version of a logging library? Or
what if they shade an incompatible logging jar within the dependency?

I don't have enough recent experience with Java build/releng practices to
know what dangers lie in wait for us, but I do have too much uncertainty
around this proposal to agree with it yet -- can you help me understand?

**Comment by ph** [comment]
I would believe core should take care of any logging library needed, but we have another problem. There is no mechanism in place to day in logstash to correctly deal with dependency of the jars. (its a bit of a far west) Plugins or core can ship with duplicates version of the jars, like the jackson library.

> For example, what if a plugin or dependency needs a conflicting version of a logging library? Or
what if they shade an incompatible logging jar within the dependency?


I am not too worried about this currently for the following reason:

- We control most of the plugin
- Good library are good citizens when using logging, like netty or kafka.

**Comment by fbaligand** [comment]
Concerning logging from third party libraries/frameworks :
- In java world, most of libraries/frameworks don't directly use a logging framework implementation (log4j, JUL, logback), but rather a logging framework abstraction (slf4j or commons-logging).
- The aim is to give applications using it, freedom to choose their logging framework implementation.
- If the application doesn't provide the bridge, logs from these libraries will be lost
- Few ones use directly log4j 1 or JUL, and since no configuration is given for these logging implementations, logs are lost or send to "undetermined" location (which is close to "lost")
- That's why at least slf4j and commons-logging bridges at least should be included in logstash. 
- Optionally, we could add log4j1.2-api and JUL bridges to log4j2

Concerning JAR conflict and shade :
- all is a question of classloader and classpath
- when java is looking for a class (for example `org.slf4j.LoggerFactory`), java runtime will search in each classloader, in a defined order, and for each classloader, in its classpath. As soon as it has found the class, it stops and don't care if another version exist elsewhere in the classpath.
- so if logstash provides these JARs in some location which is looked at first by java classloader, it ensures that they win.

**Comment by jordansissel** [comment]
> most of libraries/frameworks don't directly use a logging framework

Citation?

> so if logstash provides these JARs in some location which is looked at first by java classloader, it ensures that they win

This isn't a win. Dependency management is hard, and all versions are not completely compatible. 

----

It might be easier for me to understand and accept the proposal if we had some specific examples where we would ship a basically-unused-by-logstash-core logging library jar with Logstash releases. Can you help me with that?

Honestly, I feel like we shouldn't provide these things. Instead, we should just shade everything to isolate incompatible logging implementations (version mismatches, etc).

**Comment by fbaligand** [comment]
No citation. Just a known good practice.  
You can see here an abstract list of java libraries doing that : http://www.slf4j.org  

Then, I agree that classloading and dependency management is not an easy subject.  

Now, to speak about concrete things, after a quick review, I realize that in logstash 5.0.0-alpha5 package, there are :
- `commons-logging-1.2.jar` (dependency from manticore) and `slf4j-api-1.7.5.jar`
- `slf4j-simple-1.7.5.jar` (which is a slf4j bridge which logs directly to console at INFO level)
- `log4j-1.2.17.jar` (dependency from log4j input plugin)
- `log4j-1.2-api-2.6.1.jar`, `log4j-api-2.6.1.jar`, `log4j-core-2.6.1.jar` and `log4j-slf4j-impl-2.6.1.jar` (dependency from beats input plugin)

Thus, kafka and elasticsearch_java input/output plugins both depend on `log4j-1.2.17.jar`

So I think we have here a subject to think about :)  

A question : do you know if each logstash plugin has its own classloader ? Do you know what is classloader hierarchy in logstash/jruby ?  
It's important to know if logstash and its plugins share the same classloader or not.

**Comment by fbaligand** [comment]
Hi @jordansissel,

I made some code to understand how jruby/logstash classloaders are organized.
I added some code in log4j input plugin to display classloader hierarchy and classpath associated to each classloader.

Here is the result, from the lower classloader to the root classloader :
```
JRUBY classLoader: org.jruby.util.JRubyClassLoader
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/concurrent-ruby-1.0.0-java/lib/concurrent_ruby_ext.jar
	file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/logstash-input-beats-3.1.0.beta4-java/vendor/jar-dependencies/org/bouncycastle/bcpkix-jdk15on/1.54/bcpkix-jdk15on-1.54.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/logstash-input-beats-3.1.0.beta4-java/vendor/jar-dependencies/org/bouncycastle/bcprov-jdk15on/1.54/bcprov-jdk15on-1.54.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/jruby/lib/ruby/shared/jopenssl.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/jrjackson-0.3.9-java/lib/com/fasterxml/jackson/core/jackson-core/2.7.1/jackson-core-2.7.1.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/jrjackson-0.3.9-java/lib/com/fasterxml/jackson/module/jackson-module-afterburner/2.7.1/jackson-module-afterburner-2.7.1.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/jrjackson-0.3.9-java/lib/com/fasterxml/jackson/core/jackson-annotations/2.7.1/jackson-annotations-2.7.1.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/jrjackson-0.3.9-java/lib/com/fasterxml/jackson/core/jackson-databind/2.7.1/jackson-databind-2.7.1.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/jrjackson-0.3.9-java/lib/jrjackson/jars/jrjackson-1.2.18.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/logstash-core-event-java/lib/logstash-core-event-java/logstash-core-event-java.jar
    file:C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/jrmonitor-0.4.2/jruby-ext/target/classes/
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/puma-2.16.0-java/lib/puma/puma_http11.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/jruby-stdin-channel-0.2.0-java/lib/jruby_stdin_channel/jruby_stdin_channel.jar
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/bundle/jruby/1.9/gems/logstash-input-log4j-3.0.2-java/vendor/jar-dependencies/runtime-jars/log4j-1.2.17.jar
SYSTEM classLoader: sun.misc.Launcher$AppClassLoader
    file:/C:/Developpement/outils/logstash/test-logstash/
JRE EXT classLoader: sun.misc.Launcher$ExtClassLoader
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/access-bridge-64.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/cldrdata.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/dnsns.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/jaccess.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/jfxrt.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/localedata.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/nashorn.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/sunec.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/sunjce_provider.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/sunmscapi.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/sunpkcs11.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/ext/zipfs.jar
BOOTSTRAP classpath:
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/resources.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/rt.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/sunrsasign.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/jsse.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/jce.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/charsets.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/lib/jfr.jar
    file:/C:/Developpement/jdk/jdk1.8.0_66/jre/classes
    file:/C:/Developpement/outils/logstash/logstash-5.0.0-alpha5/vendor/jruby/lib/jruby.jar
```

**Comment by fbaligand** [comment]
So, it seems that : 
- there is only one classloader for jruby/logstash, so logstash and its plugin share the same classloader. Other are standard JVM classloaders.
- when using logstash plugins that don't depend on JARs, there is no log4j/slf4j/commons-logging JAR loaded
- when log4j input plugin is declared in my logstash configuration, log4j 1.2 JAR is added at the end of JRuby classloader classpath

That said, there are 2 possible strategies :
- either update each plugin `gemspec` to depend on the right JARs 
- either update logstash-core dependencies to add JARs slf4j, commons-logging and log4j-api-1.2 bridge

Finally, after thinking, I think first one is the cleaner strategy.

The main plugins I detect to be concerned by `gemspec` update are :
- kafka input/output plugin
- elasticsearch_java output plugin
- log4j input plugin

If this strategy is chosen, it means also to remove ` runtime 'org.apache.logging.log4j:log4j-1.2-api:2.6.2'` from here :  
https://github.com/elastic/logstash/blob/master/logstash-core/build.gradle

**Comment by talevy** [comment]
> either update each plugin gemspec to depend on the right JARs

this is not maintainable when trying to support any plugin in the ecosystem that we do not have control over.


I am leaning towards each plugin having to maintain its own set of compatibility. Even greater of an issue than just the logging jars, we have this problem across all java-jar-dependent plugins (and more and more are coming). We haven't finalized how we are to deal with the classpath conflicts that are occurring there as well.

So far, we have been going with the first strategy that you listed, but that is only because we have full control over the plugins we package with Logstash.

**Comment by fbaligand** [comment]
Each strategy has its advantages and disadvantages.

First strategy lets to add only the right JAR needed for one plugin.
But for each new plugin, it needs to pay attention to that.
And, as classpath is shared between logstash and its plugin, it means that first plugin which added slf4j JAR on classpath (for example), will win against other plugins... Not really nice...

On the other hand, second strategy let us deal with the problem only one time, and is safer with new/unknown plugins.
But it implies to add jars at logstash runtime that are not necessary in numerous cases.

Given first strategy disadvantages, it is safer to prefer second strategy as you say.

@jordansissel @ph @suyograo 
Any other opinion on that ?

**Comment by talevy** [comment]
pinging the team again here, would love to get your opinions!

**Comment by jordansissel** [comment]
I believe in Fabien's proposal and yield to greater experience. However, I
am not yet convinced we have a problem that needs solving, though, and am
somewhat content to wait until we run into this problem?

I won't block this though, if a solution PR is provided :)

**Comment by ph** [comment]
@talevy @fbaligand the more I think about it, the more I think that the core should provide theses dependencies, lets take into consideration that we dont have a lot of java plugins (for now) and the one we currently have we control the release circle.

**Comment by talevy** [comment]
@ph actually, you're absolutely right. I was originally hesitant to include it in core because of its bulk... but now that I think about it, we deliver logstash as a package with plugins (that will need these apis).

now, I am for including these libraries in core

**Comment by fbaligand** [comment]
@jordansissel 
If nothing is done, I see two concrete problems :
- if kafka or beats or elasticsearch_java or log4j plugin JARs log an error, this log is lost (because no log4j2 bridge is provided) and logstash user is not alerted.
- if a logstash user wants to customize log level of some kafka java class, to have more info in logs (debug, perf, warnings, ...) ; if he updates `log4j2.properties`, it won't work.

**Comment by fbaligand** [comment]
Concerning updates to do, up to me, there is only `.gemspec` and `build.gradle` files where dependencies have to be updated ; in logstash-core and also in some plugins ([listed in this comment](https://github.com/elastic/logstash/issues/5781#issuecomment-244593466))

**Comment by fbaligand** [comment]
@jordansissel 
As you suggested, I just created pull request to fix this issue : #5941

**Comment by talevy** [comment]
it has been decided that this is the responsibility of each plugin. Closing. Will re-open if any future issues arise
